### PR TITLE
Enable `stdenv` customization through `(buildPython*.override { inherit stdenv; })`

### PR DESCRIPTION
**Pull request by ShamrockLee**
## Description of changes

Now that https://github.com/NixOS/nixpkgs/pull/366593 has restored the missing `.override` to `buildPythonPackage` and `buildPythonApplication`, this allow specifying the `stdenv` used by the build helpers through `buildPython*.override` (as most build helpers do) instead of passing it directly into `buildPython*`.

The reason to move to `buildPythonPackage.override { stdenv = ...; }` from `buildPythonPackage { stdenv = ...; }` includes:

1. The `buildPython* { inherit stdenv; ...; }` will be impossible/unfeasible when taking fixed-point arguments (`buildPython* (finalAttrs: { ...; })`). `lib.extendMkDerivation` is on its way, but we need a clean ground to integrate it into `buildPython*`.

2. `stdenv` is a dependency `buildPythonPackage` would (directly or indirectly) get from `pkgs` through `callPackage`. It makes better sense to change it through the `.override` attribute provided by `callPackage`.

3. In the future, we could get rid of all the special, impassable arguments, and pass everything properly either into `stdenv.mkDerivation` or through `passthru`. This way, we only need `overrideAttrs` to override attributes, without having to defined special overriders like `overridePythonAttrs` and for each language and framework. `stdenv` could neither be passed into `stdenv.mkDerivation` or through `passthru`, since `stdenv.mkDerivaiton` will add a `stdenv` attribute to the resulting derivation.

For package definition that used to be:

```nix
{
  lib,
  stdenv,
  buildPythonPackage,
  fetchPyPi,
}:
buildPythonPackage rec {
  inherit stdenv;
  pname = "...";
  version = "...";
  src = fetchPypi {
    inherit pname source;
    hash = "...";
  };
}
```

can now become

```nix
{
  lib,
  stdenv,
  buildPythonPackage,
  fetchPyPi,
}:
buildPythonPackage.override { inherit stdenv; } rec {
  pname = "...";
  version = "...";
  src = fetchPypi {
    inherit pname source;
    hash = "...";
  };
}
```

In the future, when `buildPythonPackage` also takes fixed-point arguments, we could write:

```nix
{
  lib,
  stdenv,
  buildPythonPackage,
  fetchPyPi,
}:
buildPythonPackage.override { inherit stdenv; } (finalAttrs: {
  pname = "...";
  version = "...";
  src = fetchPypi {
    inherit (finalAttrs) pname source;
    hash = "...";
  };
})
```

~~This PR depends on #366593.~~
Initial effort for #271387.

## Things done

- Built on platform(s)
  - [X] x86_64-linux (no rebuilds other than `nixos-install-tools`)
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [X] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/interpreters/python/python-packages-base.nix [hunk 1]
```diff
@@ -32,20 +32,56 @@ let
...
+
   mkPythonDerivation = if python.isPy3k then
     ./mk-python-derivation.nix
   else
     ./python2/mk-python-derivation.nix;
 
-  buildPythonPackage = makeOverridablePythonPackage (lib.makeOverridable (callPackage mkPythonDerivation {
+  # This ensures that the function argument of buildPython* is always mirrored
+  # and that buildPython*.override is always preserved.
+  decorateBuildPyhon = f: lib.pipe f (map preserveFunctionOverride [
```
Comment (Steinhagen) [inline pkgs/development/interpreters/python/python-packages-base.nix:61]: I suppose this is a typo:
decorateBuildPyhon ->  decorateBuildPy**t**hon

Comment (ShamrockLee) [reply pkgs/development/interpreters/python/python-packages-base.nix]: Ahh! You're right. Thanks!

**Comment by ShamrockLee** [comment]
Bump onto `master`.

**Comment by ShamrockLee** [comment]
Merge conflicts resolved.

**Comment by philiptaron** [comment]
I should have time to review this soon.

@ShamrockLee:

1. What other PRs should I take a look at first?
2. What other PRs should I take a look at after this?

**Comment by ShamrockLee** [comment]
> 1. What other PRs should I take a look at first?

- #366593

This PR includes the commit of #366593, so merging this one effectively surpases it. Still, #366593 would be easier to merge, as it doesn't contain the interface change introduced in this PR.

> 2. What other PRs should I take a look at after this?

For fixed-point arguments support:

- #267296
- #271387

File: pkgs/development/interpreters/python/python-packages-base.nix [hunk 2]
```diff
@@ -38,7 +38,10 @@ let
         }
       else
         result
-    );
+    )
+    // lib.optionalAttrs (f ? override) {
+      override = lib.mirrorFunctionArgs f.override (fdrv: makeOverridablePythonPackage (f.override fdrv));
+    };
```
[the remark this answers is not in the document]
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/python-packages-base.nix]: Addressed and rebased.

File: pkgs/development/interpreters/python/python-packages-base.nix [hunk 3]
```diff
@@ -45,21 +45,42 @@ let
       override = lib.mirrorFunctionArgs f.override (fdrv: makeOverridablePythonPackage (f.override fdrv));
     };
 
+  overrideStdenvCompat =
+    f:
+    lib.setFunctionArgs (
+      args:
+      if args ? stdenv then
+        # TODO: warn that `args ? stdenv` is deprecated (once in-tree usage is migrated)
+        f.override { stdenv = args.stdenv; } args
```
[the remark this answers is not in the document]
Comment (ShamrockLee) [reply pkgs/development/interpreters/python/python-packages-base.nix]: Cool!

A quick grep with

```bash
find -type f -name "*.nix" \
  -exec grep -q buildPython {} \; \
  -exec grep -q "stdenv =" {} \; \
  -print
```

shows 14 relevant results.

It should be an easy tree-wide migration.

File: doc/languages-frameworks/python.section.md
[the remark this answers is not in the document]
Comment (ShamrockLee) [reply doc/languages-frameworks/python.section.md]: Should the release note entry be under "Breaking Changes" or "Other Notable Changes"?

The implementation itself is backward compatible (at least before Nixpkgs 26.05 considering the added warning), but the documentation declared the deprecation.

File: pkgs/development/python-modules/torch/source/default.nix [hunk 4]
```diff
@@ -275,14 +275,12 @@ let
...
   stdenv' = if cudaSupport then cudaPackages.backendStdenv else stdenv;
 in
-buildPythonPackage rec {
+buildPythonPackage.override { stdenv = stdenv'; } rec {
   pname = "torch";
   # Don't forget to update torch-bin to the same version.
   version = "2.8.0";
   pyproject = true;
 
-  stdenv = stdenv';
```
Comment (MattSturgeon) [inline pkgs/development/python-modules/torch/source/default.nix]: Interestingly, this seems to have triggered a rebuild:


<details> <summary>Changed packages (4)</summary>

- [nixpkgs-manual](https://search.nixos.org/packages?channel=unstable&show=nixpkgs-manual&from=0&size=50&sort=relevance&type=packages&query=nixpkgs-manual)
- [opensplatWithCuda](https://search.nixos.org/packages?channel=unstable&show=opensplatWithCuda&from=0&size=50&sort=relevance&type=packages&query=opensplatWithCuda)
- [python312Packages.torchWithCuda](https://search.nixos.org/packages?channel=unstable&show=python312Packages.torchWithCuda&from=0&size=50&sort=relevance&type=packages&query=python312Packages.torchWithCuda)
- [python313Packages.torchWithCuda](https://search.nixos.org/packages?channel=unstable&show=python313Packages.torchWithCuda&from=0&size=50&sort=relevance&type=packages&query=python313Packages.torchWithCuda)

</details>

Looks like `stdenv` was being accessed via `rec` before, so other uses need to be updated to explicitly use `stdenv'`, e.g.:

- https://github.com/ShamrockLee/nixpkgs/blob/54170644c9a252b078994022a4a8a7063c4d51b9/pkgs/development/python-modules/torch/source/default.nix#L321
- https://github.com/ShamrockLee/nixpkgs/blob/54170644c9a252b078994022a4a8a7063c4d51b9/pkgs/development/python-modules/torch/source/default.nix#L374
- https://github.com/ShamrockLee/nixpkgs/blob/54170644c9a252b078994022a4a8a7063c4d51b9/pkgs/development/python-modules/torch/source/default.nix#L458
- ...etc

Comment (ShamrockLee) [reply pkgs/development/python-modules/torch/source/default.nix]: I tried to add the `python3Packages.torch` migration as an example, but the fact that `python3Packages.torchWithCuda` gets rebuild reminds me that removing the `stdenv` argument changes the derivation arguments even though the semantics are the same.

I'll place `stdenv` into `excludeDrvArgNames` in a subsequent PR and do tree-wide migration after said PR is merged. This way we could keep this PR and the migration PRs rebuild-free. I'll also drop the `python3Packages.torch` migration for now.

File: doc/languages-frameworks/python.section.md
[the remark this answers is not in the document]
Comment (ShamrockLee) [reply doc/languages-frameworks/python.section.md]: Just realized that the release people postponed the deadline for "all breaking changes."

While we still have a remaining two-day window to merge this, I'll forward this change to #444721 and see if we could make it a release blocker.

[the remark this answers is not in the document]
Comment (MattSturgeon) [reply doc/languages-frameworks/python.section.md]: Hm, good point. I hadn't considered this from the 25.11 release window perspective.

I guess while it is a breaking change from a release-note perspective, it's also backwards compatible due to the compatibility layer... So maybe we can have our cake and eat it here; I'm not sure we need to consider this a breaking change from the release's perspective or make it a release blocker, but I'll defer to someone more knowledgeable on that.

File: pkgs/development/interpreters/python/python-packages-base.nix [hunk 5]
```diff
@@ -45,21 +45,44 @@ let
       override = lib.mirrorFunctionArgs f.override (fdrv: makeOverridablePythonPackage (f.override fdrv));
     };
 
+  overrideStdenvCompat =
+    f:
+    lib.setFunctionArgs (
+      args:
+      if args ? stdenv then
+        lib.warnIf (lib.oldestSupportedReleaseIsAtLeast 2511) ''
```
Comment (MattSturgeon) [inline pkgs/development/interpreters/python/python-packages-base.nix]: Re commit message:

> On Nixpkgs >= 25.11,
> warn about the use of buildPython\*'s deprecated argument `stdenv`.

That's not _quite_ right. The warning won't show up on 25.11+; it'll show up on 26.05+. But not until midway through the 26.05 unstable cycle, when 25.05 gets dropped and 25.11 gets marked as the oldest supported release.

That's because "oldest supported release" is technically unrelated to "what is the current unstable/release".

If we predicate the warning on `oldestSupportedReleaseIsAtLeast 2511` then 25.11 will never print the warning (intended).

---

I would rephrase to either:

> Once Nixpkgs 25.11 is the oldest supported release,

or

> On Nixpkgs >= 26.05,

The former is more accurate, if overly technical. The latter  is simpler, but could be slightly confusing during the initial part of the 26.05 cycle, when 25.05 is still the oldest supported release.

Comment (ShamrockLee) [reply pkgs/development/interpreters/python/python-packages-base.nix]: > That's not _quite_ right.

Oops!

> I would rephrase to either:
> 
> > Once Nixpkgs 25.11 is the oldest supported release,
> 
> or
> 
> > On Nixpkgs >= 26.05,
> 
> The former is more accurate, if overly technical. The latter is simpler, but could be slightly confusing during the initial part of the 26.05 cycle, when 25.05 is still the oldest supported release.

IIRC, we used to call it "After Nixpkgs 25.11 branch-off". Is it suitable here?

Comment (ShamrockLee) [reply pkgs/development/interpreters/python/python-packages-base.nix]: I see.

**Review by MattSturgeon — APPROVED** [review]
LGTM
